keps/sig-scheduling/201900715-node-teardown-pressure.md

```diff
@@ -0,0 +1,173 @@
+
+This KEP proposes a teardown pressure condition that is triggered when the termination load reaches a critical threshold signaling the scheduler to spare the node any additional workload until the backlog is resolved.
+
+Additionally, to help distribute load better, a TeardownPriority classifier is suggested, that negatively weighs the number of 
+containers under removal into the node weight.
+
+## Motivation
+
+While trying to stabilize a very particular workload, we where able to find a corner case in the kubernetes scheduler that can cause runaway overload of one or more nodes, seriously impacting the cluster performance.
+The corner case very much depends on the specific hardware setup and is not always easy triggerable, but in general results in a buildup of containers in termination state on one or multiple nodes until they become unresponsive.
+
+  1. After starting of the container, download a very large archive with 10ks of files.
+  2. Do something very minor, short lived
+  3. Exit
+
+The container then enters the termination phase in which docker removes the files from the overlay filesystem. If the removal process is to slow to finish before next container enters the termination phase, a buildup of those containers can occure.
+We found while stressing a [elasticio] workload, that some nodes accumulated 70 containers in termination state, rendering them unusable. Kubernetes-scheduler still scheduled the pods on this node while other nodes had very little load.
+
+This condition happens more often if the filesystem docker is running on is network based.
+
+[elasticio]: https://elastic.io
+
+### Goals
+
+Prevent buildup of containers in termination process and more equally distribution of load in corner cases.
+
+## Proposal
+
+We propose the "Teardown Pressure" condition as a node state that can be set by the kubelet when a threshold of containers in termination state is reached. Unlike other conditions, this will not cause the eviction manager to trigger
+pod removal, this condition only notifies the scheduler that the node should no receive any additional workload until the backlog is resolved. Additionally the kubelet should report the number of containers under removal in the RuntimeStats so the scheduler can make a more informed decision.
+
+* The RuntimeStatus message between container-manager and kublet gets enhanced by a ContainersUnderRemoval integer.
+* Kubelet gets enhanced by MaxTeardown config variable which will trigger the `NodeConditionType "TeardownPressure"` if the runtime state exceeds the threshold.
+* Kubernetes-Scheduler respects the `TeardownPressure` condition in the sense that no new work is scheduled on the node until the flag condition is `false` again.
```

Comment:
The scheduler is moving away from processing node conditions. Instead, node conditions cause the node to be tainted. The scheduler looks at node taints and if a pod does not tolerate the taint, it will not schedule the pod on that node. This node condition should follow the same approach and taint the node. In that case, this feature will be a node feature and no further changes will be needed on the scheduler side.
You should probably ask SIG Node to review this.

Reply:
We tried to use only the TeardownPressure condition to prevent further scheduling, but found that short lived containers that a scheduled by labels for example, will cause very unbalanced workload.
We have 3 nodes in our test cluster which have a label to run the batch jobs. The scheduler will start putting jobs on the first node, they then terminate which will take 1-2 minutes per container. The scheduler will then think, the node is free, but N teardown jobs are still running on the first node while the scheduler thinks, the first node has no load while the other nodes are happily idle. The scheduler then puts more jobs on the first node until the Teardown Pressure is reached and only then start putting jobs on the second node.
The result is that jobs take longer then needed, the nodes are unevenly used and resources wasted.
By putting a slight penalty for each tear-down in progress, the scheduler starts to properly distribute load according to actual machine states and not assumed load.

Reply:
I don't think it makes sense for the scheduler to blindly assume all tear-downs take a certain amount of time. This would reduce cluster resource utilization and would increase scheduling latency. Your jobs are probably more complex that take long to tear-down. Most processes take a small amount of time to tear-down. If some nodes shouldn't be used for scheduling for some amount of time, they should be tainted.

Reply:
It does not assume such a thing, it regularly gets the teardown counter from the cri backend. To crash your cluster, just create lots of files in your container before teardown. You are telling me, that because kubernetes does not care about nodes health and overload, I should implement a custom health check that regularly monitors the docker health and if the node gets overloaded, I set the uncordon for some time until the situation gets better ?